# Practical exam

In [1]:
### Please state your name here:
'Alexandra Samoila'

'Alexandra Samoila'

In [2]:
# make sure requirements are up to date
%pip install --upgrade pip >> results/requirements_log.txt
%pip install -r ../requirements.txt >> results/requirements_log.txt

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


### 1.

Given the below snippet, add the necessary code lines in function f2 in order to obtain the expected output:

In [58]:
def f2(n):
    f1 = "{'"
    f2 = "}"
    f3 = "'"
    for i in range(1, n):
        print(f"~~~ {i}")
        print(f1 + f"cubed{f3} : {i*i*i}"+ f2)
        print(f1 + f"squared{f3} : {i*i}" + f2)
        print(f"@@@ {i}")




n = 5
for x in f2(n):
    print(x)

~~~ 1
cubed' : 1
squared' : 1
@@@ 1
~~~ 2
cubed' : 8
squared' : 4
@@@ 2
~~~ 3
cubed' : 27
squared' : 9
@@@ 3
~~~ 4
cubed' : 64
squared' : 16
@@@ 4


TypeError: 'NoneType' object is not iterable

Expected output:

```
~~~ 1
{'cubed': 1}
{'squared': 1}
@@@ 1
~~~ 2
{'cubed': 8}
{'squared': 4}
@@@ 2
~~~ 3
{'cubed': 27}
{'squared': 9}
@@@ 3
~~~ 4
{'cubed': 64}
{'squared': 16}
@@@ 4
```

### 2.

Under **resources**, there are two input files, [input_file_1.csv](/2022/python_workshop/notebooks/resources/input_file_1.csv) and [input_file_2.csv](/2022/python_workshop/notebooks/resources/input_file_2.csv).

Given the two input files, remove the duplicates.

If rows in all files are combined, there must be NO duplicates. Solution should include 2 functions, one for reading data from files and one for writing data to files. A partial solution without functions is accepted.

Expected output: two separate output files under results/expected, [output_file_1.csv](/2022/python_workshop/notebooks/results/expected/output_file_1.csv) and [output_file_2.csv](/2022/python_workshop/notebooks/results/expected/output_file_2.csv).

In [18]:
import pandas as pd
test1 = pd.read_csv('./resources/input_file_1.csv', index_col = 0)
test2 = pd.read_csv('./resources/input_file_2.csv', index_col = 0)

test1.drop_duplicates(inplace= True)
test2.drop_duplicates(inplace = True)

print(test1)
print(test2)


           lastname   city
firstname                 
joe        robinson   cluj
adam         fisher  sibiu
           lastname   city
firstname                 
joe        robinson  sibiu
alex         murphy   cluj
adam         fisher  sibiu
joe        robinson   cluj


### 3.

Using the [data.gov.ro](https://data.gov.ro/dataset/acreditarea-unitatilor-sanitare-2021) API, extract the "accredited hospitals 2021" data set (link available below).

Alternatively, if the API is down, use the backup file **hospitals.csv** from the **resources** folder as a source.

In [19]:
import requests
import pandas as pd

In [43]:
from ast import arg


url_hospitals = 'https://data.gov.ro/dataset/4197b53e-7c91-4fcc-be07-883076d40ffc/resource/7a343719-d625-4b5c-98b5-895a684d61c4/download/anmcs-acreditare-unitati-sanitare-dec2021.json'


hospitals = {
    'nr. crt' : [],
    'nume' : [],
    'cod a.n.m.c.s' : [],
    'judet' :[],
    'clasificare' : [],
    'tip ups patologie' :[],
    'tip usp regim' :[],
    'tip usp invatamant' : [],
    'adresa' : [],
    'email' : [],
    'website' : [],
    'autoritate tutelara' : [],
    'acreditare cI' : [],
    'acreditare cII' : []
}

# r= requests.get(url_hospitals)
# #print(r)
# #print("HTML:\n", r.text)
# j = r.json
# #print(j)
# for item in j['nr.crt']['nume'] : 
#     try: 
#         hospitals['nr. crt'].append(int(item('Nr. Crt')))
#         hospitals['nume'].append(int(item('Nume')))
#         hospitals['cod a.n.m.c.s'].append(int(item('Cod A.N.M.C.S')))
#         hospitals['judet'].append(int(item('Judet')))
#         hospitals['clasificare'].append(int(item("Clasificare")))
#         hospitals['tip usp patologie'].append(int(item("TIP USP (in functie de specificul patologiei tratate ")))
#         hospitals['tip usp regim'].append(int(item("TIP USP (in functie de regimul proprietatii)")))
#         hospitals['tip usp invatamant'].append(int(item("TIP USP (din punct de vedere al invatamantului si al cercetarii stiintifice medicale)")))
#         hospitals['adresa'].append(int(item("Adresa completa a sediului principal")))
#         hospitals['email'].append(int(item("Adresa email oficiala")))
#         hospitals['website'].append(int(item('Website')))
#         hospitals['autoritate tutelara'].append(int(item('Autoritate tutelara')))
#         hospitals['acreditare cI'].append(int(item('Acreditare ciclul I')))
#         hospitals['acreditare cII'].append(int(item('Acreditare ciclul II')))

    
#     except Exception as e:
#         print(f'Error adding item: {item}')
#         print(e)
# hospitals = pd.DataFrame(hospitals)
#hospitals.head()

hospitals = pd.read_csv('./resources/hospital.csv', index_col = 0)
hospitals

Nr. Crt.                                               Nume  \
0          1.0                         SPITALUL MUNICIPAL CALAFAT   
1          2.0        SPITALUL DE PSIHIATRIE CRONICI SCHITU GRECI   
2          3.0                     SPITALUL CLINIC NICOLAE MALAXA   
3          4.0  SPITALUL MILITAR DE URGENTA "REGINA MARIA" BRA?OV   
4          5.0                            SPITALUL ORASENESC INEU   
...        ...                                                ...   
1360       NaN                                                NaN   
1361       NaN                                                NaN   
1362       NaN                                                NaN   
1363       NaN                                                NaN   
1364       NaN                                                NaN   

     Cod A.N.M.C.S.      Judet Clasificare  \
0              A001       Dolj          IV   
1              A002        Olt           V   
2              A003  Bucuresti         III   
3              A004     Brasov         III   
4              A005       Arad          IV   
...             ...        ...         ...   
1360            NaN        NaN         NaN   
1361            NaN        NaN         NaN   
1362            NaN        NaN         NaN   
1363            NaN        NaN         NaN   
1364            NaN        NaN         NaN   

     TIP USP (in functie de specificul patologiei tratate)  \
0                                               general      
1                                     afectiuni cronice      
2                                               general      
3                                               urgenta      
4                                               general      
...                                                 ...      
1360                                                NaN      
1361                                                NaN      
1362                                                NaN      
1363                                                NaN      
1364                                                NaN      

     TIP USP (in functie de regimul proprietatii)  \
0                                          public   
1                                          public   
2                                          public   
3                                          public   
4                                          public   
...                                           ...   
1360                                          NaN   
1361                                          NaN   
1362                                          NaN   
1363                                          NaN   
1364                                          NaN   

     TIP USP (din punct de vedere al invatamantului si al cercetarii stiintifice medicale)  \
0                                       fara invatamant                                      
1                                       fara invatamant                                      
2                         clinic cu sectii universitare                                      
3                                       fara invatamant                                      
4                                       fara invatamant                                      
...                                                 ...                                      
1360                                                NaN                                      
1361                                                NaN                                      
1362                                                NaN                                      
1363                                                NaN                                      
1364                                                NaN                                      

                   Adresa completa a sediului principal  \
0     B-dul. T. Vladimirescu, Nr. 24, Calafat, cod 2..

How many rows are in this dataset?

In [31]:
len(hospitals)

1365

What are the columns in this dataset?

In [32]:
hospitals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1365 entries, 0 to 1364
Data columns (total 17 columns):
 #   Column                                                                                 Non-Null Count  Dtype  
---  ------                                                                                 --------------  -----  
 0   Nr. Crt.                                                                               762 non-null    float64
 1   Nume                                                                                   762 non-null    object 
 2   Cod A.N.M.C.S.                                                                         762 non-null    object 
 3   Judet                                                                                  762 non-null    object 
 4   Clasificare                                                                            760 non-null    object 
 5   TIP USP (in functie de specificul patologiei tratate)                       

Drop the unnecessary columns:
* Nr. Crt.
* TIP USP (din punct de vedere al invatamantului si al cercetarii stiintifice medicale)
* the last three columns

In [44]:
#d_columns = ['Nr. Crt', 'TIP USP (din punct de vedere al invatamantului si al cercetarii stiintifice medicale)']
hospitals.drop(hospitals.columns[[0,7,14,15,16]], inplace = True, axis = 1 )
hospitals.head()

Nume Cod A.N.M.C.S.  \
0                         SPITALUL MUNICIPAL CALAFAT           A001   
1        SPITALUL DE PSIHIATRIE CRONICI SCHITU GRECI           A002   
2                     SPITALUL CLINIC NICOLAE MALAXA           A003   
3  SPITALUL MILITAR DE URGENTA "REGINA MARIA" BRA?OV           A004   
4                            SPITALUL ORASENESC INEU           A005   

       Judet Clasificare  \
0       Dolj          IV   
1        Olt           V   
2  Bucuresti         III   
3     Brasov         III   
4       Arad          IV   

  TIP USP (in functie de specificul patologiei tratate)  \
0                                            general      
1                                  afectiuni cronice      
2                                            general      
3                                            urgenta      
4                                            general      

  TIP USP (in functie de regimul proprietatii)  \
0                                       public   
1                                       public   
2                                       public   
3                                       public   
4                                       public   

                Adresa completa a sediului principal  \
0  B-dul. T. Vladimirescu, Nr. 24, Calafat, cod 2...   
1                     SLATINA STR. A. I. CUZA NR. 14   
2  STR. DIMITRIE CANTEMIR, NR.1, PARTER, SECT.4, ...   
3  Bucuresti, Str. Institutul Medico-Militar, Nr....   
4                            Calea Republicii, nr. 5   

         Adresa email oficiala                   Website  \
0      spitalcalafat@gmail.com          spitalcalafat.ro   
1     spitalulschitu@gmail.com    www.spitalul-schitu.ro   
2  secretariat@spitalmalaxa.ro       www.spitalmalaxa.ro   
3      smureginamaria@rdsbv.ro          www.smubrasov.ro   
4          spitalineu@yahoo.ro  http://www.spitalineu.ro   

                                Autoritatea tutelara Acreditare ciclul I  \
0                      PRIMARIA MUNICIPIULUI CALAFAT           Acreditat   
1                             CONSILIUL JUDETEAN OLT           Acreditat   
2                                              ASSMB           Acreditat   
3  DIRECtIA MEDICALa a Ministerului Apararii Nati...           Acreditat   
4                             Primaria Orasului Ineu           Acreditat   

                                Acreditare ciclul II  
0  Categoria V Decizie de prelungire a procesului...  
1  Categoria V Decizie de prelungire a procesului...  
2         Categoria IV Acreditat cu incredere redusa  
3              Categoria II Acreditat cu recomandari  
4  Categoria V Decizie de prelungire a procesului...

Apply the following transformations to the columns, so that they are easier to work with:
* lowercase all columns
* `TIP USP (in functie de specificul patologiei tratate)` rename to `tip_patologie`
* `TIP USP (in functie de regimul proprietatii)` rename to `tip`
* `Adresa completa a sediului principal` rename to `adresa`
* `Adresa email oficiala` rename to `email`
* `Autoritatea tutelara` rename to `autoritate_tutelara`
* `Acreditare ciclul I` rename to `acreditare_i`
* `Acreditare ciclul II` rename to `acreditare_ii`

In [46]:
hospitals = hospitals.rename(columns=lambda x: x.lower())
#hospitals.head()

hospitals.rename(
    columns={
        'tip usp (in functie de specificul patologiei tratate)': 'tip_patologie',
        'tip usp (in functie de regimul proprietatii)' : 'tip',
        'adresa completa a sediului principal' : 'adresa',
        'adresa email oficiala' : 'email',
        'autoritatea tutelara'  : 'autoritate_tutelara',
        'acreditare ciclul i' : 'acreditare_i',
        'acreditare ciclul ii' : 'acreditare_ii'

        
    },
    inplace=True # careful with this!
)
hospitals.columns

Index(['nume', 'cod a.n.m.c.s.', 'judet', 'clasificare', 'tip_patologie',
       'tip', 'adresa', 'email', 'website', 'autoritate_tutelara',
       'acreditare_i', 'acreditare_ii'],
      dtype='object')

The total number of hospitals in this dataset is 762, however the actual number of rows is greater than that.

Identify and remove the empty (blank space) rows.

In [49]:
hospitals.dropna(inplace = True)
hospitals.head()

nume cod a.n.m.c.s.  \
0                         SPITALUL MUNICIPAL CALAFAT           A001   
1        SPITALUL DE PSIHIATRIE CRONICI SCHITU GRECI           A002   
2                     SPITALUL CLINIC NICOLAE MALAXA           A003   
3  SPITALUL MILITAR DE URGENTA "REGINA MARIA" BRA?OV           A004   
4                            SPITALUL ORASENESC INEU           A005   

       judet clasificare      tip_patologie     tip  \
0       Dolj          IV            general  public   
1        Olt           V  afectiuni cronice  public   
2  Bucuresti         III            general  public   
3     Brasov         III            urgenta  public   
4       Arad          IV            general  public   

                                              adresa  \
0  B-dul. T. Vladimirescu, Nr. 24, Calafat, cod 2...   
1                     SLATINA STR. A. I. CUZA NR. 14   
2  STR. DIMITRIE CANTEMIR, NR.1, PARTER, SECT.4, ...   
3  Bucuresti, Str. Institutul Medico-Militar, Nr....   
4                            Calea Republicii, nr. 5   

                         email                   website  \
0      spitalcalafat@gmail.com          spitalcalafat.ro   
1     spitalulschitu@gmail.com    www.spitalul-schitu.ro   
2  secretariat@spitalmalaxa.ro       www.spitalmalaxa.ro   
3      smureginamaria@rdsbv.ro          www.smubrasov.ro   
4          spitalineu@yahoo.ro  http://www.spitalineu.ro   

                                 autoritate_tutelara acreditare_i  \
0                      PRIMARIA MUNICIPIULUI CALAFAT    Acreditat   
1                             CONSILIUL JUDETEAN OLT    Acreditat   
2                                              ASSMB    Acreditat   
3  DIRECtIA MEDICALa a Ministerului Apararii Nati...    Acreditat   
4                             Primaria Orasului Ineu    Acreditat   

                                       acreditare_ii  
0  Categoria V Decizie de prelungire a procesului...  
1  Categoria V Decizie de prelungire a procesului...  
2         Categoria IV Acreditat cu incredere redusa  
3              Categoria II Acreditat cu recomandari  
4  Categoria V Decizie de prelungire a procesului...

How many public hospitals and how many private hospitals are in this dataset?

In [50]:
hospitals.tip.value_counts()

public    380
privat    375
Name: tip, dtype: int64

Inspect the values in the `acreditare_i` column + print them.

In [55]:
hospitals['acreditare_i'].info()
res = hospitals.drop_duplicates(subset = ["acreditare_i"])
print(res)

<class 'pandas.core.series.Series'>
Int64Index: 755 entries, 0 to 761
Series name: acreditare_i
Non-Null Count  Dtype 
--------------  ----- 
755 non-null    object
dtypes: object(1)
memory usage: 11.8+ KB
                                      nume cod a.n.m.c.s.  judet  \
0               SPITALUL MUNICIPAL CALAFAT           A001   Dolj   
188               SOCIETATEA DR IRIMIA SRL         A50009  Arges   
424  S.C. MUNTENIA MEDICAL COMPETENCES S.A           B001  Arges   

      clasificare tip_patologie     tip  \
0              IV       general  public   
188  Neclasificat       general  privat   
424             V       general  privat   

                                                adresa  \
0    B-dul. T. Vladimirescu, Nr. 24, Calafat, cod 2...   
188                    PITESTI, str.Negru Voda, nr.3-5   
424   Arges, Pitesti, Str. Pictor N Grigorescu nr. 2 A   

                          email                  website  \
0       spitalcalafat@gmail.com         spitalcalafat.r

Create a dataframe containing the hospitals that are *not* accredited.

How many hospitals are *not* accredited?

Pick a county (judet) for a more in-depth analysis.
(Bucuresti is also a county)

In [56]:
judet = 'Bucuresti'

According to this dataset, how many hospitals are in this county?

In [57]:
hospitals.judet.value_counts()

Bucuresti          103
Dolj                47
Iasi                42
Constanta           42
Cluj                36
Prahova             35
Brasov              30
Timis               30
Bacau               26
Bihor               25
Arges               22
Mures               21
Suceava             19
Maramures           17
Buzau               17
Sibiu               16
Neamt               16
Hunedoara           14
Valcea              13
Alba                13
Arad                13
Galati              12
Gorj                11
Olt                 10
Vrancea              9
Teleorman            9
Calarasi             8
Mehedinti            8
Ilfov                8
Botosani             8
Caras-Severin        7
Vaslui               7
Salaj                7
Harghita             7
Braila               7
Satu Mare            7
Dambovita            6
Tulcea               6
Bistrita-Nasaud      6
Covasna              5
Giurgiu              5
Ialomita             5
Name: judet, dtype: int64

For the chosen county, create a Dataframe that contains the following information:
* judet
* tip (public, privat)
* count of unaccredited hospitals
* count of accredited hospitals

In [64]:
stat = hospitals[['judet','tip','acreditare_i']].copy()
stat.head()

judet     tip acreditare_i
0       Dolj  public    Acreditat
1        Olt  public    Acreditat
2  Bucuresti  public    Acreditat
3     Brasov  public    Acreditat
4       Arad  public    Acreditat

Create a plot that highlights
* the number of accredited public hospitals
* the number of accredited private hospitals
* the number of unaccredited public hospitals
* the number of unaccredited private hospitals

The expected output for Bucuresti can be found  [here](/2022/python_workshop/notebooks/results/expected/output_test.png).

In [ ]:
plot_acreditat = ...

In [ ]:
plot_neacreditat = ...

In [ ]:
plot_data = ...
plot_data.head()

In [ ]:
plot_data.plot(kind="bar")